lib/hw_id/hw_id.c

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <hw_id.h>
+#include <zephyr/kernel.h>
+
+#define IMEI_LEN 15
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_BLE_MAC)
+/* Ask BLE stack for default MAC address */
+
+#include <bluetooth/bluetooth.h>
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+	struct bt_le_oob oob;
+	int ret = bt_le_oob_get_local(BT_ID_DEFAULT, &oob);
+
+	if (ret) {
+		return -EIO;
+	}
+
+	snprintk(buf, buf_len, "%02X%02X%02X%02X%02X%02X",
+		oob.addr.a.val[5],
+		oob.addr.a.val[4],
+		oob.addr.a.val[3],
+		oob.addr.a.val[2],
+		oob.addr.a.val[1],
+		oob.addr.a.val[0]);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_BLE_MAC) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_DEVICE_ID)
+/* Directly read Device ID from registers */
+
+#include <zephyr/drivers/hwinfo.h>
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+
+	uint8_t device_id[8];
+	ssize_t ret = hwinfo_get_device_id(device_id, ARRAY_SIZE(device_id));
+
+	if (ret != ARRAY_SIZE(device_id)) {
+		return -EIO;
+	}
+
+	snprintk(buf, buf_len, "%02X%02X%02X%02X%02X%02X%02X%02X",
+		device_id[0],
+		device_id[1],
+		device_id[2],
+		device_id[3],
+		device_id[4],
+		device_id[5],
+		device_id[6],
+		device_id[7]);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_DEVICE_ID) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_UUID)
+/* Request a UUID from the modem */
+
+#include <modem/modem_jwt.h>
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+
+	struct nrf_device_uuid dev = {0};
+
+	int err = modem_jwt_get_uuids(&dev, NULL);
+
+	if (err) {
+		return -EIO;
+	}
+
+	snprintk(buf, buf_len, "%s", dev.str);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_UUID) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_IMEI)
+/* Ask the modem for an IMEI */
+
+#include <nrf_modem_at.h>
+int hw_id_get(char *buf, size_t buf_len)
+{
+	char imei_buf[IMEI_LEN + 6 + 1]; /* Add 6 for \r\nOK\r\n and 1 for \0 */
+
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+
+	/* Retrieve device IMEI from modem. */
+	int err = nrf_modem_at_cmd(imei_buf, ARRAY_SIZE(imei_buf), "AT+CGSN");
+
+	if (err) {
+		return -EIO;
+	}
+
+	/* Set null character at the end of the device IMEI. */
+	imei_buf[IMEI_LEN] = 0;
+
+	snprintk(buf, buf_len, "%s", imei_buf);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_IMEI) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_NET_MAC)
+/* Use MAC address from default network interface */
+
+#include <zephyr/net/net_if.h>
+int hw_id_get(char *buf, size_t buf_len)
```

Comment:
Maybe an idea would be to use the same function entry for all sources and then have individual internal functions that are called depending on the source. This way you would get rid of the duplication of the input variable checks.

Reply:
Although this reduces duplication of actual code, I'd have to copy the `#if defined` stuff twice. I'd like to skip that indirection for the moment.